src/Middleware/MethodNotAllowedMiddleware.php

```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-router for the canonical source repository
+ * @copyright Copyright (c) 2018 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive-router/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Router\Middleware;
+
+use Fig\Http\Message\StatusCodeInterface as StatusCode;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Webimpress\HttpMiddlewareCompatibility\HandlerInterface as RequestHandlerInterface;
+use Webimpress\HttpMiddlewareCompatibility\MiddlewareInterface;
+use Zend\Expressive\Router\RouteResult;
+
+use const Webimpress\HttpMiddlewareCompatibility\HANDLER_METHOD;
+
+/**
+ * Emit a 405 Method Not Allowed response
+ *
+ * If the request composes a route result, and the route result represents a
+ * failure due to request method, this middleware will emit a 405 response,
+ * along with an Allow header indicating allowed methods, as reported by the
+ * route result.
+ *
+ * If no route result is composed, and/or it's not the result of a method
+ * failure, it passes handling to the provided handler.
+ */
+class MethodNotAllowedMiddleware implements MiddlewareInterface
+{
+    /**
+     * @var ResponseInterface
+     */
+    private $response;
+
+    public function __construct(ResponseInterface $response)
```

Comment:
I know that everywhere in v2 we pass response prototype, not a factory. Isn't it an issue as we experienced in v3?
In v3 this become a callback, so this is BC break in v3.

Reply:
If you change it in v2, it would cause a BC break as well and thus would require a new major version according to semver.

Reply:
This is new class in v2, so maybe for new classes we should have factories (callbacks) and all these cases should be somehow handled in factories (that case: `MethodNotAllowedMiddlewareFactory`).

Reply:
I'm removing this class entirely. The only other classes that have a response prototype then are the `ImplicitHeadMiddleware` and the `RouteMiddleware`, which is consistent with what is in zend-expressive v2 currently. Additionally, both of those have BC breaks for v3 regardless; the former swaps its response for a router, and the latter removes the response entirely.